Improve .onLoad() with system.file() for robust Java initialization

cofinoa opened the pull request:
### Summary

This PR improves `.onLoad()` to:
- Use `system.file()` for robust Java directory resolution.
- Check JVM initialization status (`.jniInitialized`) before setting `java.parameters`.
- Warn the user if the JVM is already initialized, to avoid silent failures.

This ensures better compatibility across installed and development modes and avoids JVM conflicts.

Please review and let me know if tests or documentation updates are needed.

cofinoa committed:
Update onLoad.R

cofinoa committed:
Refactor JVM memory setting in .onLoad function to check initialization status

jbedia commented:
Excellent, thanks for this improvement! One question: are we sure most users will have 2Gb of heap space available? What if not? Would the `"-Xmx2g"` request fail silently?

cofinoa commented:
Yes! You’re right, the `-Xmx2g` setting assumes users have at least 2 GB of available heap space.

If the system cannot allocate that much, the JVM typically falls back silently to the maximum it can allocate, but this behaviour can depend on the Java version and environment.

To make this more robust, we could:

1. Allow users to override the memory setting via an R option or environment variable.
2. Add a small runtime check to log the actual allocated memory (Runtime.getMaxMemory()).

I would recommend opening a follow-up issue for configurability and memory reporting, rather than adding it to this PR.

cofinoa commented:
Yes! You’re right, the `-Xmx2g` setting assumes users have at least 2 GB of available heap space.

If the system cannot allocate that much, the JVM typically falls back silently to the maximum it can allocate, but this behaviour can depend on the Java version and environment.

To make this more robust, we could:

1. Allow users to override the memory setting via an R option or environment variable.
2. Add a small runtime check to log the actual allocated memory (Runtime.getMaxMemory()).

I have edited the issue #14 to follow-up this for configurability and memory reporting, rather than adding it to this PR.

Commit:
Enhance .onLoad(): configurable JVM memory and test coverage

cofinoa commented:
@mor427unican 

I would like access to the JAR MANIFEST.MF of the `netcdf-java` library which is loaded in `loadeR.java`, to extract some of the metadata, and print on the message, like the version and/or build date, for example. 

I can do it, in a java main class, like iterating over all Class Loader of the JVM, but in loadeR.java I have not suceeded:

```java
// Java Main class to iterate over all class loaders and print MANIFEST.MF
import java.io.InputStream;
import java.util.jar.Manifest;
import java.util.jar.Attributes;

public class ManifestFinder {
    public static void main(String[] args) {
        ClassLoader cl = Thread.currentThread().getContextClassLoader();
        while (cl != null) {
            System.out.println("Checking ClassLoader: " + cl);
            try (InputStream stream = cl.getResourceAsStream("META-INF/MANIFEST.MF")) {
                if (stream != null) {
                    Manifest manifest = new Manifest(stream);
                    Attributes attr = manifest.getMainAttributes();
                    for (Object keyObj : attr.keySet()) {
                        String key = keyObj.toString();
                        String value = attr.getValue(key);
                        System.out.println(key + ": " + value);
                    }
                } else {
                    System.out.println("No MANIFEST.MF found in this ClassLoader");
                }
            } catch (Exception e) {
                System.out.println("Error reading MANIFEST.MF: " + e.getMessage());
            }
            cl = cl.getParent();
        }
    }
}

please take a look. 

I also would suggest to use its own class loader for loadeR.java, and use it for every call to the `netcdf-java`

This may be, needs a different branc/issue and/or PR

jbedia commented:
These are all extremely useful enhancements, thanks for that!
Another question: Could we execute the `Runtime.getMaxMemory()` on package load and print on screen this info? 

A pseudocode example follows:

```
packageStartupMessage("The maximum JVM heap space available is:",  (Runtime.getRuntime().maxMemory() / (1024 * 1024)), " MB");
packageStartupMessage("Available memory in the current session can be configured using options(java.parameters = \"-Xmx8g\")"
```

Commit:
Available heap space and jar metadata at load time

mor427unican commented:
I made some changes to the `loadeR.java` package (on the following branch: `cofinoa-feature/improve-onload-java-init`). Specifically, I addressed issue #14 by adding support for configuring the Java heap space before loading the package with `library(loadeR.java)`.

For example, users can do (this is also shown in the README):

```r
# Set an R option
options(loadeR.java.memory = "-Xmx4g")
library(loadeR.java)

# Or use an environment variable
Sys.setenv(JAVA_TOOL_OPTIONS = "-Xmx4g")
library(loadeR.java)
```
In addition, when loading the package with `library(loadeR.java)`, the following information is now displayed:

1. The amount of Java heap space available (either user-defined or default if none was set).
2.  Metadata from the JAR used by loadeR.java, extracted from the MANIFEST.MF file (e.g., version, build date, etc.).

cofinoa committed:
Update DESCRIPTION and onLoad functions to suggest testthat and improve JVM memory handling

cofinoa commented:
@mor427unican I have updated both `.onLoad` and `.onAttach` :

- .onLoad: Now properly handles java.parameters as a vector. It checks if -Xmx is present and replaces it with -Xmx2g if its value is less than 2g. If -Xmx is missing, it is added to the vector while preserving other parameters. The parameters are correctly separated and updated using options(java.parameters).

- .onAttach: moved here the reports the maximum heap memory available to the JVM, and properties of the MANIFEST, has been parsed and used in the message. We need to review this, because I was looking to use the classLoader, and NO to open the JAR and read the manifest. It should be possible using the class loader.

Please review and test the changes.

jbedia commented:
Hi!
just tested the new features in my local. Everything looks fine, very nice!
```R
> library(loadeR)
Loading required package: rJava
Loading required package: loadeR.java
Java version 11x amd64 by Ubuntu detected
The maximum JVM heap space available is: 2.00 GB
NetCDF Java Library Version: 4.6.0-SNAPSHOT (Built-On: 20150423.1338) loaded and ready
Loading required package: climate4R.UDG
climate4R.UDG version 0.2.6 (2023-06-26) is loaded
Please use 'citation("climate4R.UDG")' to cite this package.
loadeR version 1.8.2 (2024-06-04) is loaded
Please use 'citation("loadeR")' to cite this package.
```

jbedia reviewed (approved):
Well done, thanks!

cofinoa commented:
FYI: 

I have opened an PR s-u/rJava#341 in the `rJava` repo raising the issue of having the default parameter `-Xmx512m`.

Commit:
Use of class loader (.onAttach) and jinit with JAR classpath (.onLoad)

mor427unican commented:
I have reviewed and updated both `.onAttach` and `.onLoad` as requested (on the following branch: `cofinoa-feature/improve-onload-java-init`):

* `.onAttach`: The code now uses the `ClassLoader` of `ucar.nc2.NetcdfFile` to access the `META-INF/MANIFEST.MF` resource. This avoids directly opening the JAR file and follows the approach requested.
* `.onLoad`: I replaced the previous use of `.jpackage()` with an explicit call to `.jinit(classpath = ...)`, to ensure the JVM is initialized with the correct classpath before `.onAttach()` runs. This guarantees that classes from the JAR (e.g., `ucar.nc2.NetcdfFile`) are available when needed.

Here is a short demonstration:

```r
> options(java.parameters = "-Xmx4g") 
> library(loadeR.java)
Cargando paquete requerido: rJava
Java version 23x amd64 by Azul Systems, Inc. detected
The maximum JVM heap space available is: 4.00 GB
NetCDF Java Library Version: 4.6.0-SNAPSHOT (Built-On: 20150423.1338) loaded and ready
```

cofinoa commented:
@jbedia we are planning a "new" release (v1.1.2)  with the same version of netcdf-java as v1.1.1 and functionality, but adding the Unit Tests to this version, to have these tests running, to update the netcdf-java and use them as tests for the next release. I don't have a clear visions of those changes will be major, changes (i.e. v1.2.1) or major (v2.0), probablly the last, because I foreseen some refactoring and therefore we will move some funcitionality around (i.e. the plan is to moe Java calls to loadeR.java, and create just R wrappers functions, and uses them form loadeR)

@mor427unican for releasing the next version of loadeR.java, v1.1.2, do you have a clear picture of what is still missing?.

mor427unican commented:
For the upcoming `loadeR.java` release, I believe the changes related to MANIFEST.MF extraction via the ClassLoader are now in place and stable. I have also implemented tests for the functions within `loadeR.java`. As for the `loadeR` package, I have written tests for 27 R files so far. Within those 27, the remaining functions that still require suitable datasets to be properly tested are: 
* `getMemberDomain()` 
* `getRunTimeDomain.decadal()`
* `getRunTimeDomain.seasonal()`
* `getForecastTimeDomain()`
* `dataInventory.NetCDF.ts()`

mariadlor committed:
Update test-loadeRjava.R
